syncapi/synctypes/clientevent.go

```diff
@@ -78,63 +80,62 @@ func ToClientEvents(serverEvs []gomatrixserverlib.PDU, format ClientEventFormat,
 		if se == nil {
 			continue // TODO: shouldn't happen?
 		}
-		if format == FormatSyncFederation {
-			evs = append(evs, ToClientEvent(se, format, string(se.SenderID()), se.StateKey(), spec.RawJSON(se.Unsigned())))
+		ev, err := ToClientEvent(se, format, userIDForSender)
+		if err != nil {
+			logrus.Errorf("Failed converting event to ClientEvent: %s", err.Error())
 			continue
 		}
+		evs = append(evs, *ev)
+	}
+	return evs
+}
 
-		sender := spec.UserID{}
-		validRoomID, err := spec.NewRoomID(se.RoomID())
+// ToClientEvent converts a single server event to a client event.
+// It provides default logic for event.SenderID & event.StateKey -> userID conversions.
+func ToClientEventDefault(userIDQuery spec.UserIDForSender, event gomatrixserverlib.PDU) ClientEvent {
+	ev, err := ToClientEvent(event, FormatAll, userIDQuery)
+	if err != nil {
+		return ClientEvent{}
+	}
+	return *ev
+}
+
+// If provided state key is a user ID (state keys beginning with @ are reserved for this purpose)
+// fetch it's associated sender ID and use that instead. Otherwise returns the same state key back.
+//
+// # This function either returns the state key that should be used, or an error
+//
+// TODO: handle failure cases better (e.g. no sender ID)
+func FromClientStateKey(roomID spec.RoomID, stateKey string, senderIDQuery spec.SenderIDForUser) (*string, error) {
+	if len(stateKey) >= 1 && stateKey[0] == '@' {
+		parsedStateKey, err := spec.NewUserID(stateKey, true)
 		if err != nil {
-			continue
-		}
-		userID, err := userIDForSender(*validRoomID, se.SenderID())
-		if err == nil && userID != nil {
-			sender = *userID
+			// If invalid user ID, then there is no associated state event.
+			return nil, fmt.Errorf("Provided state key begins with @ but is not a valid user ID: %s", err.Error())
```

Comment:
`%w` instead of `%s` please, so we can unwrap the error and use things like `errors.Is()`